typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1361 @@
+                  })))
+      t0 t1
+
+  let rec singleton path leaf =
+    match path with
+    | [] -> { usage = leaf; children = Projection.Map.empty }
+    | proj :: path ->
+        {
+          usage = Unused;
+          children = Projection.Map.singleton proj (singleton path leaf);
+        }
+end
+
+(** Lift Usage_tree to forest *)
+module Usage_forest : sig
+  module Path : sig
+    type t
+
+    val child : t -> Projection.t -> t
+    val child_of_many : t list -> Projection.t -> t list
+    val fresh_root_of_ident : Ident.t -> t
+    val fresh_root : string -> t
+  end
+
+  type t
+
+  val seq : t -> t -> t
+  val par : t -> t -> t
+  val seqs : t list -> t
+  val pars : t list -> t
+  val unused : t
+  val singleton : Path.t -> Usage.t -> t
+  val map : (Path.t -> Usage.t -> Usage.t) -> t -> t
+end = struct
+  module Root_id = struct
+    module T = struct
+      type t = { id : int; name : string }
+
+      let compare t1 t2 = t1.id - t2.id
+    end
+
+    include T
+    module Map = Map.Make (T)
+
+    let stamp = ref 0
+
+    let fresh name =
+      let id = !stamp in
+      stamp := id + 1;
+      { id; name }
+
+    let fresh_of_ident ident = fresh (Ident.name ident)
+    let name t1 = t1.name
+    let print ppf t = Format.fprintf ppf "%s" (name t)
+  end
+
+  type t = Usage_tree.t Root_id.Map.t
+  (** Represents a forest of usage. *)
+
+  let _print ppf t =
+    Root_id.Map.iter
+      (fun rootid tree ->
+        Format.fprintf ppf "%a = %a, " Root_id.print rootid Usage_tree.print
+          tree)
+      t
+
+  module Path = struct
+    type t = Root_id.t * Usage_tree.Path.t
+
+    let child ((rootid, path) : t) proj : t =
+      (rootid, Usage_tree.Path.child path proj)
+
+    let child_of_many paths proj = List.map (fun path -> child path proj) paths
+    let fresh_root name : t = (Root_id.fresh name, Usage_tree.Path.root)
+
+    let fresh_root_of_ident ident =
+      (Root_id.fresh_of_ident ident, Usage_tree.Path.root)
+  end
+
+  let unused = Root_id.Map.empty
+
+  (** [f] must be monotone  *)
+  let map2 f t0 t1 =
+    Root_id.Map.merge
+      (fun _rootid t0 t1 ->
+        match (t0, t1) with
+        | None, None -> assert false
+        | None, Some t1 -> Some t1
+        | Some t0, None -> Some t0
+        | Some t0, Some t1 -> Some (f t0 t1))
+      t0 t1
+
+  let par t0 t1 = map2 Usage_tree.par t0 t1
+  let seq t0 t1 = map2 Usage_tree.seq t0 t1
+  let pars l = List.fold_left par unused l
+  let seqs l = List.fold_left seq unused l
+
+  let singleton ((rootid, path') : Path.t) leaf =
+    Root_id.Map.singleton rootid (Usage_tree.singleton path' leaf)
+
+  (** f must be monotone *)
+  let map f =
+    Root_id.Map.mapi (fun root tree ->
+        Usage_tree.map (fun projs usage -> f (root, projs) usage) tree)
+end
+
+module UF = Usage_forest
+
+module Ienv : sig
+  type t
+
+  val par : t -> t -> t
+  val seq : t -> t -> t
+  val pars : t list -> t
+  val seqs : t list -> t
```

Comment:
I think it is a bit confusing using the `seq` and `par` names for these, since they don't quite correspond to the same thing here as they do for usages. I would suggest renaming `seq` as `concat` and renaming `par` as something like `meet`.

Reply:
I now call them `Ienv.Extension.conjunct` and `disjunct`. Does that sound OK?